File: c02_4_maximum_subarray_sum.py
import unittest

def algorithm1(array):
    n=len(array)
    best=0
    for a in range(n):
        for b in range(a,n):
            sum_=0
            for k in array[a:b+1]:
                sum_+=k
            best=max(best,sum_)
    return best

def algorithm2(array):
    n=len(array)
    best=0
    for a in range(n):
        sum_=0
        for b in array[a:]:
            sum_+=b
            best=max(best,sum_)
    return best

def algorithm3(array):
    best,sum_=0,0
    for k in array:
        sum_=max(k,sum_+k)
        best=max(best,sum_)
    return best

class Test(unittest.TestCase):
    def test(self):
        for a in (algorithm1,algorithm2,algorithm3):
            with self.subTest(a=a):
                self.assertEqual(a([]),0)
                self.assertEqual(a([-1]),0)
                self.assertEqual(a([-1,2,4,-3,5,2,-5,2]),10)

if __name__=='__main__':
    unittest.main()

File: c++/c02_4_maximum_subarray_sum.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	ll algorithm1(const vector<ll>& array) {
		ll n = ssize(array);
		ll best = 0;
		for (ll a = 0; a < n; ++a) {
			for (ll b = a; b < n; ++b) {
				ll sum = 0;
				for (ll k = a; k <= b; ++k) {
					sum += array[k];
				}
				best = max(best, sum);
			}
		}
		return best;
	}

	ll algorithm2(const vector<ll>& array) {
		ll n = ssize(array);
		ll best = 0;
		for (ll a = 0; a < n; ++a) {
			ll sum = 0;
			for (ll b = a; b < n; ++b) {
				sum += array[b];
				best = max(best, sum);
			}
		}
		return best;
	}

	ll algorithm3(const vector<ll>& array) {
		ll n = ssize(array);
		ll best = 0, sum = 0;
		for (ll k = 0; k < n; ++k) {
			sum = max(array[k], sum + array[k]);
			best = max(best, sum);
		}
		return best;
	}

	void test(auto f) {
		EXPECT_EQ(f({}), 0);
		EXPECT_EQ(f({ -1 }), 0);
		EXPECT_EQ(f({ -1,2,4,-3,5,2,-5,2 }), 10);
	}
}

TEST(C024MaximumSubarraySum, algorithm1) {
	test(algorithm1);
}

TEST(C024MaximumSubarraySum, algorithm2) {
	test(algorithm2);
}

TEST(C024MaximumSubarraySum, algorithm3) {
	test(algorithm3);
}
